From 03d797003091de8bef3f80e7d63dcf4002ff5164 Mon Sep 17 00:00:00 2001
From: Pascal Paillet <p.paillet@foss.st.com>
Date: Wed, 27 Nov 2024 09:37:30 +0100
Subject: [PATCH 1616/1695] drivers: stm32_cpu_opp: improve SCP set voltage

Remove get voltage call. On STM32MP13, it is a lot more
efficient to set a bit in a PWR register than calling
get_voltage() as there is no cache in the regulator
framework.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 9061197911e7dc354c3490ab71f07c6190d2adf9 drivers: stm32_cpu_opp: improve SCP set voltage]
Change-Id: Ib7fd72c2fa86534b33ba293508c76b4e0fa13f7b
Signed-off-by: Pascal Paillet <p.paillet@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/439763
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Pascal PAILLET <p.paillet@st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Tested-by: Pascal PAILLET <p.paillet@st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/drivers/stm32_cpu_opp.c | 11 ++---------
 1 file changed, 2 insertions(+), 9 deletions(-)

diff --git a/core/drivers/stm32_cpu_opp.c b/core/drivers/stm32_cpu_opp.c
index ac955818e..e1f3fdb87 100644
--- a/core/drivers/stm32_cpu_opp.c
+++ b/core/drivers/stm32_cpu_opp.c
@@ -294,15 +294,9 @@ static TEE_Result scp_set_regu_voltage(struct regulator *regulator __unused,
 				       int uv)
 {
 	TEE_Result res = TEE_ERROR_GENERIC;
-	int cur_uv = 0;
-
-	cur_uv = regulator_get_voltage(cpu_opp.regul);
-	if (uv == cur_uv)
-		return TEE_SUCCESS;
 
 #ifdef CFG_STM32MP13
-	if (cur_uv > MPU_RAM_LOW_SPEED_THRESHOLD &&
-	    uv <= MPU_RAM_LOW_SPEED_THRESHOLD)
+	if (uv <= MPU_RAM_LOW_SPEED_THRESHOLD)
 		io_setbits32(stm32_pwr_base(), PWR_CR1_MPU_RAM_LOW_SPEED);
 #endif
 
@@ -311,8 +305,7 @@ static TEE_Result scp_set_regu_voltage(struct regulator *regulator __unused,
 		return res;
 
 #ifdef CFG_STM32MP13
-	if (cur_uv <= MPU_RAM_LOW_SPEED_THRESHOLD &&
-	    uv > MPU_RAM_LOW_SPEED_THRESHOLD)
+	if (uv > MPU_RAM_LOW_SPEED_THRESHOLD)
 		io_clrbits32(stm32_pwr_base(), PWR_CR1_MPU_RAM_LOW_SPEED);
 #endif
 
-- 
2.47.2

